### PR TITLE
UI: Host Volumes

### LINKED ISSUE #6362: UI support for Host Volumes
## Background

Host volumes are client-local volumes that can optionally mounted by Nomad as specified in the client config. Jobs can then be specified to require host volumes, forcing the scheduler to place the allocations for said job on clients with matching host volumes.

See how this looks in the CLI here: https://github.com/hashicorp/nomad/pull/6011
And how it looks in client and job specs here:  https://github.com/hashicorp/nomad/pull/5923

## UI Requirements

1. Show host volumes for a client on the Client Detail page, including the same information that is shown in the CLI.
2. Explore adding host volumes to faceted search on the Client List page.
3. Show host volume requirements on the Task Group page.
4. Show host volume usage on the Allocation and Task Detail pages.

## Implementation notes

We know that [Container Storage Interface support](https://github.com/hashicorp/nomad/issues/5378) is coming, so we should be mindful of this when designing changes to especially the task group/allocation/task pages. 

Host volumes is distinct from CSI, but they both relate to storage so there is bound to be some user flow overlap.

### PR DESCRIPTION
**Pull request by DingoEatingFuzz**
Show host volume information throughout the Web UI. Closes #6362 

#### This includes changes to:
1. Task group page
2. Allocation page
3. Task page
4. Client page
5. Allocation tables
6. Task tables
7. Task group tables
8. Client table

The primary benefit is you can now inspect what volumes an allocation and a task require and trace that back to the host volumes as declared in client configurations.

#### Screenshots
image
_Volume Requirements table on the task group detail page._

image
_New Volume columns in the task group table and allocation table_.

image
_New Volume column in the task table on the allocation detail page._

image
_Volumes table on the task detail page._

image
_New Volumes column in the clients table._

image
_Volumes table on the client detail page._

### CODE REVIEW

File: ui/tests/acceptance/client-detail-test.js [hunk 1]
```diff
@@ -900,6 +901,36 @@ module('Acceptance | client detail', function(hooks) {
     assert.ok(ClientDetail.eligibilityToggle.isDisabled);
     assert.ok(ClientDetail.drainPopover.isDisabled);
   });
+
+  test('the host volumes table lists all host volumes in alphabetical order by name', async function(assert) {
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/tests/acceptance/client-detail-test.js]: Ha, I must have gotten distracted between writing the test description and implementation.

File: ui/tests/acceptance/task-detail-test.js [hunk 2]
```diff
@@ -128,6 +128,44 @@ module('Acceptance | task detail', function(hooks) {
...
+  test('when a task does not have volumes, the volumes table is not shown', async function(assert) {
+    const job = server.create('job', { createAllocations: false, noHostVolumes: true });
+    allocation = server.create('allocation', { jobId: job.id, clientStatus: 'running' });
+    task = server.db.taskStates.where({ allocationId: allocation.id })[0];
+
+    await Task.visit({ id: allocation.id, name: task.name });
+    assert.notOk(Task.hasVolumes);
+  });
+
+  test('each volume in the volumes table shows information about the volume', async function(assert) {
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/tests/acceptance/task-detail-test.js]: I see how the test name is slightly inaccurate, but the `each x` pattern is in a lot of our tests so I'd rather keep this consistent.

File: ui/app/models/volume-mount.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { computed } from '@ember/object';
+import attr from 'ember-data/attr';
+import Fragment from 'ember-data-model-fragments/fragment';
+import { fragmentOwner } from 'ember-data-model-fragments/attributes';
+
+export default Fragment.extend({
+  task: fragmentOwner(),
+
+  volume: attr('string'),
+  source: computed('task.taskGroup.volumes.@each.{name,source}', function() {
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/app/models/volume-mount.js]: It shouldn't ever change, but it should also be a dependent key for Correctness™ sake.

File: ui/app/templates/allocations/allocation/task/index.hbs [hunk 4]
```diff
@@ -105,6 +105,32 @@
     </div>
   {{/if}}
 
+  {{#if model.task.volumeMounts.length}}
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/app/templates/allocations/allocation/task/index.hbs]: I guess it's just JavaScript habits making their way into Handlebars? I don't have a good reason other than that.

File: ui/app/templates/components/job-page/parts/task-groups.hbs [hunk 5]
```diff
@@ -11,6 +11,7 @@
         {{#t.sort-by prop="name"}}Name{{/t.sort-by}}
         {{#t.sort-by prop="count"}}Count{{/t.sort-by}}
         {{#t.sort-by prop="queuedOrStartingAllocs" class="is-3"}}Allocation Status{{/t.sort-by}}
+        {{#t.sort-by prop="volumes.length"}}Volume?{{/t.sort-by}}
```
[the remark this answers is not in the document]
Comment (DingoEatingFuzz) [reply ui/app/templates/components/job-page/parts/task-groups.hbs]: All the Yes/No columns should be consistent, I think. Whether or not to use a question mark I don't have a strong preference for. Do you have thoughts, @joshklekamp?

[the remark this answers is not in the document]
Comment (joshklekamp) [reply ui/app/templates/components/job-page/parts/task-groups.hbs]: I noticed that this morning as well. I'd prefer to not do a question mark. I think the original "Volumes?" was a holdover from previous mocks.

File: ui/app/templates/allocations/allocation/task/index.hbs [hunk 6]
```diff
@@ -105,6 +105,32 @@
...
+            <th>Name</th>
+            <th>Destination</th>
+            <th>Permissions</th>
+            <th>Client Source</th>
+          {{/t.head}}
+          {{#t.body as |row|}}
+            <tr data-test-volume>
+              <td data-test-volume-name>{{row.model.volume}}</td>
+              <td data-test-volume-destination><code>{{row.model.destination}}</code></td>
+              <td data-test-volume-permissions>{{if row.model.readOnly "Read" "Read/Write"}}</td>
```
Comment (joshklekamp) [inline ui/app/templates/allocations/allocation/task/index.hbs]: I know the CLI declares permissions via read_only = "true/false", but I'd be curious to see how CSI is phrasing permissions. However we're doing it, I think being consistent across host volumes and CSI, both in the CLI and the UI, is probably pretty important. I like "Read" and "Read/Write" better than the options I had in the original mocks FWIW!

Comment (DingoEatingFuzz) [reply ui/app/templates/allocations/allocation/task/index.hbs]: 👍 CSI volumes _should_ follow suit, but I'll keep in eye out.

**Pull request merged**